File: day23.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

struct Node {
  int value = 0;
  int next;  // Index in |cups| of the next node.
  int pred;  // Index in |cups| of the predecessor.

  Node (int v) : value(v) {}
};

const int N = 1e6;
const int kNumMoves = 1e7;


std::vector<Node> initialise(std::string input) {
  std::vector<Node> cups;
  const int n = input.length();
  std::unordered_map<int, int> indices;

  // Initialise value and predecessor for the input nodes.
  for (int i=0; i<n; ++i) {
    int value = input[i] - '0';
    cups.push_back(Node(value));
    indices.insert({value, i});
  }
  for (Node& v : cups) {
    if (v.value == 1) continue;
    v.pred = indices[v.value - 1];
  }

  // Initialise value and predecessor for remaining nodes.
  for (int i=n+1; i<=N; ++i) {
    Node v(i);
    if (i == n+1) v.pred = indices[n];
    else v.pred = i-2;
    cups.push_back(v);
  }
  cups[indices[1]].pred = N-1;

  // Initialise pointer to the next node.
  for (int i=0; i<N; ++i)
    cups[i].next = (i+1)%N;

  return cups;
}


int main() {
  // Initialise the linked list.
  std::string input;
  std::cin >> input;
  std::vector<Node> cups = initialise(input);

  auto value = [&](int idx) { return cups[idx].value; };
  auto next = [&](int idx) { return cups[idx].next; };
  auto pred = [&](int idx) { return cups[idx].pred; };

  // Perform moves.
  int current = 0;
  for (int i=0; i<kNumMoves; ++i) {
    // Remove the next three cups.
    int first = next(current);
    int second = next(first);
    int third = next(second);
    cups[current].next = next(third);

    // Find the destination cup.
    int dest = pred(current);
    while (value(dest) == value(first)
        || value(dest) == value(second)
        || value(dest) == value(third)) {
      dest = pred(dest);
    }

    // Insert the removed cups.
    cups[third].next = next(dest);
    cups[dest].next = first;

    current = next(current);
  }

  // Find cup 1 and print the answer.
  int one = input.find("1");
  std::cout << (long long) value(next(one))
             * (long long) value(next(next(one)));
  return 0;
}

File: day23.hs
import qualified Data.Map as M
import Data.Char (digitToInt)
import Data.List (elemIndex, sort)
import Data.Maybe (fromJust)
import Util

type IM = M.Map Int Int

initialise :: [Int] -> ([IM], Int)
initialise v = (arrs, idx1)
  where
    n = 10^6
    vals = v ++ [length v + 1 .. n]
    pred = let idx = M.fromList $ zip v [0..]
               f (val, i) = if val==1 then (i, n-1)
                            else (i, idx M.! (val-1))
               a = map snd . sort . map f $ M.toList idx
               b = (idx M.! length v):[length v .. n-2]
           in a++b
    next = [1 .. n-1] ++ [0]
    arrs = map (M.fromList . zip [0..]) [vals, pred, next]
    idx1 = fromJust $ elemIndex 1 v

perform :: IM -> IM -> (Int, IM) -> (Int, IM)
perform vals pred (current, next) = (next'' M.! current, next'')
  where
    -- Remove the next three cups.
    first = next M.! current
    second = next M.! first
    third = next M.! second
    next' = M.insert current (next M.! third) next

    -- Find the destination cup.
    getDest i xs = if (vals M.! i) `notElem` xs then i
                   else getDest (pred M.! i) xs
    dest = let vs = map (vals  M.! ) [first, second, third]
           in getDest (pred M.! current) vs

    -- Insert the removed cups.
    next'' = M.insert third (next' M.! dest)
           . M.insert dest first $ next'

solve ([vals, pred, next], idx1) = cup1 * cup2
  where moves = iterate (perform vals pred) (0, next)
        next' = snd (moves !! (10^7))
        cup1 = vals M.! (next' M.! idx1)
        cup2 = vals M.! (next' M.! (next' M.! idx1))

main = interact $ Util.print . solve . initialise . parse
  where parse = map digitToInt . head . lines
